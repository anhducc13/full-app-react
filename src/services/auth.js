import axios from 'axios';
const URL = process.env.REACT_APP_API_URL

const register = params => {
    return axios.post(`${URL}/auth/register`, params)
}

const verify = (accessToken) => {
    return axios.get(`${URL}/auth/verify/${accessToken}`)
}

const login = (params) => {
    return axios.post(`${URL}/auth/login`, params)
}

const logout = (accessToken) => {
    const reqConfig = {
        "headers": {
            "Authorization": `Bearer ${accessToken}`,
        }

    }
    return axios.get(`${URL}/auth/logout`, reqConfig)
}

const updatePassword = (params, accessToken) => {
    const reqConfig = {
        "headers": {
            "Authorization": `Bearer ${accessToken}`,
        }
    }
    return axios.post(`${URL}/auth/updatePassword`, params, reqConfig)
}

const forgotPassword = params => {
    return axios.post(`${URL}/auth/forgotPassword`, params)
}

export default {
    login,
    logout,
    register,
    verify,
    updatePassword,
    forgotPassword
}