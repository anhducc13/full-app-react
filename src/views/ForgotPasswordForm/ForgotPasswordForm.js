import React from 'react';
import Avatar from '@material-ui/core/Avatar';
import Button from '@material-ui/core/Button';
import CssBaseline from '@material-ui/core/CssBaseline';
import TextField from '@material-ui/core/TextField';
import Link from '@material-ui/core/Link';
import Grid from '@material-ui/core/Grid';
import LockOutlinedIcon from '@material-ui/icons/LockOutlined';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';
import Container from '@material-ui/core/Container';
import { Link as RouteLink} from 'react-router-dom'

const useStyles = makeStyles(theme => ({
  '@global': {
    body: {
      backgroundColor: theme.palette.common.white,
    },
  },
  paper: {
    marginTop: theme.spacing(2),
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
  },
  avatar: {
    margin: theme.spacing(1),
    backgroundColor: theme.palette.secondary.main,
    width: 60,
    height: 60,
  },
  form: {
    width: '100%', // Fix IE 11 issue.
    marginTop: theme.spacing(3),
  },
  submit: {
    margin: theme.spacing(3, 0, 2),
  },
}));

export const ForgotPasswordForm = (props) => {
  const classes = useStyles();

  const {
    emailField,
    usernameField,
  } = props.textField;

  const {
    onChangeEmail,
    onChangeUsername,
  } = props.handleChange

  const submitForm = props.handleSubmit

  return (
    <Container component="main" maxWidth="sm">
      <CssBaseline />
      <div className={classes.paper}>
        <Avatar className={classes.avatar}>
          <LockOutlinedIcon />
        </Avatar>
        <Typography component="h1" variant="h4">
          Forgot Password
        </Typography>
        <form className={classes.form} onSubmit={submitForm} noValidate>
          <TextField
            variant="outlined"
            margin="normal"
            required
            fullWidth
            id="email"
            label="Email"
            name="email"
            autoFocus
            value={emailField.value}
            error={emailField.error !== ''}
            onChange={onChangeEmail}
            helperText={emailField.valid ? "" : emailField.error}
          />
          <TextField
            variant="outlined"
            margin="normal"
            required
            fullWidth
            id="username"
            label="Username"
            name="username"
            value={usernameField.value}
            error={usernameField.error !== ''}
            onChange={onChangeUsername}
            helperText={usernameField.valid ? "" : usernameField.error}
          />
          <Button
            size="large"
            type="submit"
            fullWidth
            variant="contained"
            color="primary"
            className={classes.submit}
            disabled={
              !emailField.valid || !usernameField.valid
            }
          >
            Reset
          </Button>
          <Grid container>
            <Link to="/login" variant="body2" component={RouteLink}>
              Have an account? Login
            </Link>
          </Grid>
        </form>
      </div>
    </Container>
  );
}