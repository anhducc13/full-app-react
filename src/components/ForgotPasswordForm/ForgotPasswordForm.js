import React, { useState } from 'react';
import { setGlobal } from 'reactn';
import Avatar from '@material-ui/core/Avatar';
import CssBaseline from '@material-ui/core/CssBaseline';
import Link from '@material-ui/core/Link';
import Grid from '@material-ui/core/Grid';
import LockOutlinedIcon from '@material-ui/icons/LockOutlined';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';
import Container from '@material-ui/core/Container';
import { Link as RouteLink } from 'react-router-dom';
import { Button as ButtonCustom } from '../shared/Button';
import { InputText } from '../shared/InputText';
import Swal from 'sweetalert2';
import { authService } from 'services';
import { validateUsername, validateEmail } from 'helpers/validators';

const useStyles = makeStyles(theme => ({
  '@global': {
    body: {
      backgroundColor: theme.palette.common.white,
    },
  },
  paper: {
    marginTop: theme.spacing(2),
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
  },
  avatar: {
    margin: theme.spacing(1),
    backgroundColor: theme.palette.secondary.main,
    width: 60,
    height: 60,
  },
  form: {
    width: '100%', // Fix IE 11 issue.
    marginTop: theme.spacing(3),
  },
}));

export const ForgotPasswordForm = (props) => {
  const classes = useStyles();
  const initTextField = {
    value: '',
    error: '',
    valid: false,
  };
  const [emailField, setEmailField] = useState(initTextField);
  const [usernameField, setUsernameField] = useState(initTextField);

  const onChangeEmail = (e) => {
    let {value, error, valid} = emailField;
    value = e.target.value;
    const outputValidate = validateEmail({value, error, valid});
    setEmailField(outputValidate);
  }

  const onChangeUsername = (e) => {
    let {value, error, valid} = usernameField;
    value = e.target.value;
    const outputValidate = validateUsername({value, error, valid});
    setUsernameField(outputValidate);
  }

  const onSubmitForm = (e) => {
    e.preventDefault()
    const payload = {
      "username": usernameField.value,
      "email": emailField.value,
    }
    setGlobal({
      loading: true,
    })
    authService.forgotPassword(payload)
      .then(() => {
        setGlobal({
          loading: false,
        })
        Swal.fire(
          'Thành công!',
          'Vui lòng kiểm tra email để lấy lại mật khẩu',
          'success'
        )
        props.history.push('/')
      })
      .catch(err => {
        setGlobal({
          loading: false,
        })
        if (err.response && err.response.status === 400) {
          Swal.fire(
            'Có lỗi xảy ra!',
            'Dữ liệu đầu vào có vấn đề',
            'error'
          )
        }
        else if (err.response && err.response.status === 404) {
          Swal.fire(
            'Có lỗi xảy ra!',
            'Username hoặc email chưa được đăng ký',
            'error'
          )
        } else {
          props.history.push('/error')
        }
      })
  }

  return (
    <Container component="main" maxWidth="sm">
      <CssBaseline />
      <div className={classes.paper}>
        <Avatar className={classes.avatar}>
          <LockOutlinedIcon />
        </Avatar>
        <Typography component="h1" variant="h4">
          Quên mật khẩu
        </Typography>
        <form className={classes.form} noValidate onSubmit={onSubmitForm}>
          <InputText
            textField={emailField}
            name="email"
            label="Email"
            type="email"
            onChange={onChangeEmail}
          />
          <InputText
            textField={usernameField}
            name="username"
            label="Tên đăng nhập"
            type="text"
            onChange={onChangeUsername}
          />
          <ButtonCustom 
            displayText="Nhận mật khẩu mới"
            disabled={
              !usernameField.valid || !emailField.valid
            }
          />
          <Grid container>
            <Link to="/dang-nhap" variant="body2" component={RouteLink}>
              Đã có tài khoản? Đăng nhập
            </Link>
          </Grid>
        </form>
      </div>
    </Container>
  );
}