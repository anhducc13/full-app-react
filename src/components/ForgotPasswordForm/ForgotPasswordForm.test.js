import React from 'react';
import { mount, shallow } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { ForgotPasswordForm } from './ForgotPasswordForm';
import { authService } from '../../services';

describe('Forgot Password Form Render', () => {
  it('should render correctly', () => {
    const component = shallow(<ForgotPasswordForm />);
    expect(component).toMatchSnapshot();
  });
});

describe('Forgot Password form validate', () => {
  it('validate email', () => {
    const component = mount(
      <MemoryRouter>
        <ForgotPasswordForm />
      </MemoryRouter>
    );
    const targetInput = {
      target: {
        value: ""
      }
    }
    component.find('input[name="email"]').simulate('change', targetInput);
    expect(component.find('p#email-helper-text').text())
      .toBe('Email không hợp lệ');
    component.unmount()
  })

  it('validate username', () => {
    const component = mount(
      <MemoryRouter>
        <ForgotPasswordForm />
      </MemoryRouter>
    );
    const targetInput = {
      target: {
        value: " "
      }
    }
    component.find('input[name="username"]').simulate('change', targetInput);
    expect(component.find('p#username-helper-text').text())
      .toBe('Tên đăng nhập không được chứa khoảng trắng');
    component.unmount()
  })

  it('validate success', () => {
    const component = mount(
      <MemoryRouter>
        <ForgotPasswordForm />
      </MemoryRouter>
    );
    const data = {
        'username': 'anhducc13',
        'email': 'trantienduc10@gmail.com'
      }
    component.find('input[name="username"]').simulate('change', {target: {value: data['username']}});
    component.find('input[name="email"]').simulate('change', {target: {value: data['email']}});
    expect(component.find('p#username-helper-text').length)
      .toBe(0);
    expect(component.find('p#email-helper-text').length)
      .toBe(0);
    component.unmount()
  })
});

describe('Forgot password form submit', () => {
  it('Button submit disable with invalid data', () => {
    const component = mount(
      <MemoryRouter>
        <ForgotPasswordForm />
      </MemoryRouter>
    );
    component.find('input[name="username"]').simulate('change', { target: { value: '' } });
    component.find('input[name="email"]').simulate('change', { target: { value: 'ANhduc13' } });
    expect(component.find('button[type="submit"]').prop('disabled')).toBeTruthy();
    component.unmount();
  })

  it('Button submit enable with expect data', () => {
    const component = mount(
      <MemoryRouter>
        <ForgotPasswordForm />
      </MemoryRouter>
    );
    component.find('input[name="username"]').simulate('change', { target: { value: 'anhducc13' } });
    component.find('input[name="email"]').simulate('change', { target: { value: 'trantienduc10@gmail.com' } });
    expect(component.find('button[type="submit"]').prop('disabled')).toBeFalsy();
    component.unmount();
  })

  it('Submit form, call api with valid input', () => {
    const component = mount(
      <MemoryRouter>
        <ForgotPasswordForm />
      </MemoryRouter>
    );
    const value = {
      'username': 'anhducc13',
      'email': 'trantienduc10@gmail.com',
    }
    component.find('input[name="username"]').simulate('change', { target: { value: value['username'] } });
    component.find('input[name="email"]').simulate('change', { target: { value: value['email'] } });
    const forgorPasswordFn = jest.spyOn(authService, 'forgotPassword').mockImplementation(() => Promise.resolve({ status: 200 }));
    component.find('form').simulate('submit');
    expect(forgorPasswordFn).toHaveBeenCalled();
    expect(forgorPasswordFn).toHaveBeenCalledWith(value);
    component.unmount();
  })
});
