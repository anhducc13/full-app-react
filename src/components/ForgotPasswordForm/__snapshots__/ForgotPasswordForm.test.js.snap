// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forgot Password Form Render should render correctly 1`] = `
<WithStyles(ForwardRef(Container))
  component="main"
  maxWidth="sm"
>
  <CssBaseline />
  <div
    className="makeStyles-paper-1"
  >
    <WithStyles(ForwardRef(Avatar))
      className="makeStyles-avatar-2"
    >
      <LockOutlinedIcon />
    </WithStyles(ForwardRef(Avatar))>
    <WithStyles(ForwardRef(Typography))
      component="h1"
      variant="h4"
    >
      Quên mật khẩu
    </WithStyles(ForwardRef(Typography))>
    <form
      className="makeStyles-form-3"
      noValidate={true}
      onSubmit={[Function]}
    >
      <InputText
        label="Email"
        name="email"
        onChange={[Function]}
        textField={
          Object {
            "error": "",
            "valid": false,
            "value": "",
          }
        }
        type="email"
      />
      <InputText
        label="Tên đăng nhập"
        name="username"
        onChange={[Function]}
        textField={
          Object {
            "error": "",
            "valid": false,
            "value": "",
          }
        }
        type="text"
      />
      <Button
        disabled={true}
        displayText="Nhận mật khẩu mới"
      />
      <WithStyles(ForwardRef(Grid))
        container={true}
      >
        <WithStyles(ForwardRef(Link))
          component={[Function]}
          to="/dang-nhap"
          variant="body2"
        >
          Đã có tài khoản? Đăng nhập
        </WithStyles(ForwardRef(Link))>
      </WithStyles(ForwardRef(Grid))>
    </form>
  </div>
</WithStyles(ForwardRef(Container))>
`;
